### ISSUE TITLE
Level 4 Lock()

### ISSUE THREAD

**Issue by yongchein-o** [nus-cs2030/2122-s2#54]
# Summary
Since a locked question inherits from an unlocked question, why can't the compiler find answer()

## Description 
In the question we have the following: 

```
jshell> Question q = new FillInBlank("Snow white and the ? dwarfs", 7)
q ==> Snow white and the ? dwarfs; Your answer: 0

jshell> q.answer(3).answer(7).lock().answer(5)
|  Error:
|  cannot find symbol
|    symbol:   method answer(int)
|  q.answer(3).answer(7).lock().answer(5)
|  ^---------------------------------^

jshell> q.answer(3).answer(7).lock() instanceof FillInBlank
$.. ==> true

```

From this, we know that lock() will return a lockedfillinblank as it is an instance of FillInBlank. But why is it that we cannot call answer() using lockedfillinblank. 

I have a class

`class LockedFillInBlank extends FillInBlank implements Lock`

## Idea s
Describe the approach used (if applicable)

## Screenshots (if any): 
Insert Images here if necessary

**Comment by jingyulim1** [comment]
From the other test cases, we see that lock() can lock other types of question as well. So lock() should return a lockedquestion which does not have answer(). Then, due to compile time, the error will appear.

**Comment by bryankwe** [comment]
@jingyulim1 thanks for replying my question in the other thread as well

I understand what you are saying, but from the above "instanceof" test case, doesn't this mean that LockedQuestion has to extend FillInBlank? 

In this case (and I guess what @yongchein-o  is asking, is, even if you do not specify an answer() method, wouldn't it just call the parent's (FillInBlank) answer() method in the end?

**Comment by jingyulim1** [comment]
It won't, if LockedQuestion is implemented as an interface/abstract class (depending on how you approach it) that the 3 types of question can implement/extends.

**Comment by yongchein-o** [comment]
My LockedFillInBlank is a **concrete** class that extends FillInBlank but yet has no ability to call answer() in  FillInBlank

**Comment by jingyulim1** [comment]
If you return it as a Lock type( Lock lock()), then it would still face compile time error since answer() is not declared in Lock. 

That is, a Lock object doesn't know answer(), so it can't compile.

**Comment by jingyulim1** [comment]
If you return it as a Lock type (Lock lock()), then it would still face compile time error.  But if you return as a LockedFillInBlank type(LockedFillInBlank lock()) then I think it should be able to.

**Comment by haoyangw** [comment]
Yeah if you let LockedFillInBlank extend FillInBlank then LockedFillInBlank will inherit answer() and not give you the 'cannot find symbol' errors you want(assuming your answer() function is default/protected/public access).

Have you considered the opposite inheritance relationship? If you let FillInBlank extend LockedFillInBlank, and you don't define answer() in LockedFillInBlank, then what happens when you do a `LockedFillInBlank q = new FillInBlank(...)`? Also, in this case, would `q instanceof FillInBlank` return true?

**Comment by hongyaong** [comment]
Hi all, thanks for the tips. I still have an error for level4 whereby q, cannot find method mark(). I'm wondering if it is because Question class don't have mark() method that is why there is this error. I placed my mark() method in my Locked interface is this the wrong way to do it?

**Comment by haoyangw** [comment]
> I'm wondering if it is because Question class don't have mark() method that is why there is this error

Yeah correct, and that's some for the test cases. Isn't q.mark() supposed to throw a cannot find symbol error?

> I placed my mark() method in my Locked interface is this the wrong way to do it?
I think this logic is not wrong, I wrote my code this way as well. Perhaps you set the return type for the lock() method wrongly? Because the return type for that method has to define the mark() method

**Comment by Jingting203** [comment]
Hi I'm still very confused over the order of inheritance. If you let FillInBlank extend LockedFillInBlank, where does the Question class go? I assumed that FillInBlank would extend the Question class

**Comment by loqir** [comment]
hi i understand that we can return a locked interface with a mark() method upon calling the method lock(), but how can we make Question initially be an unlocked interface such that it cannot access the method mark()?

Also in the unlocked interface we have an answer() method with a return type of Question but since Question implements the unlock interface, does this cause cyclic dependency?

**Comment by jlee1007** [comment]
I don't think there's a need to create a LockedFilledInBlank class. Rather, you can just create 2 interfaces Question & LockedQuestion with the different methods declared in each. The 3 qn types can inherit both. So based on the methods defined in each, we can control the access. e.g. if mark() is only in LockedQn, you won't be able to mark an unlocked question

**Comment by loqir** [comment]
does an abstract class work like an interface in the sense that the interface segregration principle can also be applied to an abstract class?

**Comment by loqir** [comment]
ahh i see thanks for the detailed explanation!
